shapely/coordinates.py

```diff
@@ -54,20 +69,252 @@ def transform(geometry, transformation, include_z=False):
     if new_coordinates.dtype != np.float64:
         raise ValueError(
             "The provided transformation returned an array with an unexpected "
-            f"dtype ({new_coordinates.dtype})"
+            f"dtype ({new_coordinates.dtype}, but expected {coordinates.dtype})"
         )
     if new_coordinates.shape != coordinates.shape:
         # if the shape is too small we will get a segfault
         raise ValueError(
             "The provided transformation returned an array with an unexpected "
-            f"shape ({new_coordinates.shape})"
+            f"shape ({new_coordinates.shape}, but expected {coordinates.shape})"
         )
     geometry_arr = lib.set_coordinates(geometry_arr, new_coordinates)
     if geometry_arr.ndim == 0 and not isinstance(geometry, np.ndarray):
         return geometry_arr.item()
     return geometry_arr
 
 
+def transform_extra(
```

Comment:
It could already came up in discussions with @jorisvandenbossche , but I couldn't find it:  what is the reason of adding a separate function for the two extra parameters? Wouldn't `transform_extra(geometry, transformation, interleaved=False, rebuild=False)` be equivalent to `transform(geometry, transformation)`?

Reply:
> It could already came up in discussions with @jorisvandenbossche , but I couldn't find it: what is the reason of adding a separate function for the two extra parameters? Wouldn't `transform_extra(geometry, transformation, interleaved=False, rebuild=False)` be equivalent to `transform(geometry, transformation)`?

Hi @caspervdw thanks for your review!
`transform_extra(geometry, transformation, interleaved=True, rebuild=False)` is equivalent to `transform(geometry, transformation)` (I'm not a big fan of the keyword `interleaved`, so maybe if it confused you we should consider alternatives) 
Could you please clarify what are you suggesting? Are you suggesting to add the extra keywords to the original `transform` function?

Reply:
My main reasoning behind not modifying `transform`, making `transform_planar` public and not adding a common function like `_run_maybe_vectorized_transformation` was performance, as I wanted to reduce the function calls and unnecessary checks for some common use cases.

Reply:
Hi @idanmiara, thank you for your recent efforts on shapely!

Why are you worried about performance? Looking at the code, I don't expect any significant difference between the existing `transform` and the `transform_extra` (provided that `rebuild=False`). 

Either way, a comprehensible and maintainable API is also worth a great deal.  So yes I'd prefer a single top-level function `transform` that has all the options. But I may be convinced by some benchmarks :)

Reply:
Hi @caspervdw,
I think I've addressed all the issues, please review.
I left the extra 4 top level functions (for whoever want to choose a specific variations for performance reasons) I would run some benchmarks to justify this reasoning and we can discuss if it makes sense to hide them.